File: dopamine/packages/core/src/effects/confetti.ts
/**
 * Confetti as an `EffectFactory` on the Dopamine backbone.
 *
 * The quintessential celebration: a burst of paper confetti POPS upward from the
 * action then TUMBLES DOWN under gravity with air-drag flutter — spinning
 * rectangles + petals in many OKLCH hues that sway, settle, and fade. The
 * signature is the downward, physical, fluttering fall, deliberately distinct
 * from Solarbloom's UPWARD drifting motes.
 *
 * Like the other built-ins it is DATA + a-few-lines: the mood→params mapping +
 * OKLCH palette live in confetti.dope.json (loader-resolved). ALL renderer
 * plumbing — program/VAO, standard uniforms, the light+shadow loop, dispose — is
 * the shared `createPassInstance` generic fullscreen-pass runner. The only code
 * here is the confetti SHADER + a small config naming its scalar params, its
 * shadow height, and the per-frame launch-then-fall amplitude envelope.
 */

import { CONFETTI_FRAGMENT_SRC, CONFETTI_VERTEX_SRC, MAX_PIECES } from "../engine/confetti-shader.js";
import { easeOutBack, easeOutCubic } from "../engine/tempo.js";
import type { EffectContext, EffectFactory, EffectInstance, FeelingInput } from "../framework/effect.js";
import { registerEffect } from "../framework/registry.js";
import { registerProgram } from "../framework/programs.js";
import { parseDope, resolveDopeParams } from "../framework/loader.js";
import { createPassInstance, type PassConfig, type PassParams } from "../framework/pass-runner.js";
import doc from "./confetti.dope.json";

const DOPE = parseDope(doc as object);

interface ConfettiParams extends PassParams {
  exposure: number;
  pieceCount: number;
  spread: number;
  launchSpeed: number;
  gravity: number;
  flutter: number;
  pieceSize: number;
  spin: number;
  overshoot: number;
  pieceSeed: number;
}

/**
 * Launch-then-fall amplitude over normalized life. Unlike Solarbloom's
 * held-breath `envelope` (which decays from its early peak), confetti must stay
 * BRIGHT through the long fall — per-piece `particleFade` in the shader handles
 * each piece dimming as it lands. So this is a sharp POP attack (overshoot at
 * launch), a near-full sustain while everything falls, then a gentle fade only at
 * the very end as the last pieces settle. Peak > 1 at the launch pop.
 */
function confettiAmp(life: number, overshoot: number): number {
  if (life <= 0 || life >= 1) return 0;
  const attack = 0.12;
  if (life < attack) {
    // Sharp pop with a little overshoot (the burst leaving the action).
    return easeOutBack(life / attack, overshoot);
  }
  // Long luminous sustain, then a soft fade over the last ~30% as pieces settle.
  const tailStart = 0.7;
  if (life < tailStart) return 1;
  const x = (life - tailStart) / (1 - tailStart);
  return 1 - easeOutCubic(x) * 0.85;
}

// The launched + falling cloud spans a good chunk of the viewport; give the
// shadow a moderate occluder "height".
const SHADOW_HEIGHT_FRAC = 0.5;

const CONFIG: PassConfig = {
  vertex: CONFETTI_VERTEX_SRC,
  fragment: CONFETTI_FRAGMENT_SRC,
  uniforms: [
    "uExposure", "uPieceCount", "uSpread", "uLaunchSpeed", "uGravity",
    "uFlutter", "uPieceSize", "uSpin", "uPieceSeed",
  ],
  usesOrigin: true,
  // overshoot feeds the envelope, not a uniform; pieceSeed is the scatter offset.
  bindings: { overshoot: null, pieceSeed: "uPieceSeed" },
  shadowHeightFrac: SHADOW_HEIGHT_FRAC,
  frame: ({ life }, params) => ({
    amp: confettiAmp(life, (params as ConfettiParams).overshoot),
  }),
};

function createInstance(params: ConfettiParams, ctx: EffectContext): EffectInstance {
  return createPassInstance(CONFIG, params as PassParams, ctx);
}

export const confetti: EffectFactory<ConfettiParams> = {
  name: "confetti",
  resolve: (feeling: FeelingInput) =>
    resolveDopeParams(DOPE, feeling, { MAX_PIECES }, "pieceSeed") as unknown as ConfettiParams,
  create: createInstance,
  reducedMotion: { peakMs: 320, holdMs: 420 },
};

// Expose the renderer as a bundled PROGRAM so `loadEffect()` can bind an
// arbitrary host-authored `.dope` (one that references program "confetti") to it
// with no code. Purely numeric/palette params — no code-shaped composition.
registerProgram<ConfettiParams>("confetti", {
  create: createInstance,
  scatterKey: "pieceSeed",
  consts: { MAX_PIECES },
  reducedMotion: { peakMs: 320, holdMs: 420 },
});

export default registerEffect(confetti);

File: effects/confetti/android/Confetti.kt
// Confetti as a Dopamine effect on the Android backbone — mirror of the web
// `effect-confetti/src/index.ts` + swift's `Confetti.swift`.
//
// The quintessential celebration: a burst of paper confetti POPS upward from the
// action then TUMBLES DOWN under gravity with air-drag flutter — spinning
// rectangles + petals in many OKLCH hues. Per the generalization mandate, the
// ONLY per-effect code is {the GLSL shader + the bespoke tempo + a tiny config
// naming its scalar params / bindings / shadow height / the per-frame
// launch-then-fall amplitude}. Everything else — the `.dope` mapping, the OKLCH
// palette, the registry, the fullscreen-pass runner, the standard uniforms — is
// shared backbone. The numeric/palette bag comes verbatim from the bundled
// `.dope` (the SAME bytes as the web), resolved by the shared loader (byte-parity
// proven). Confetti is PURE-SHADER (no panel): `createPassInstance` + `PassConfig`.

package ai.dopamine.effect.confetti

import ai.dopamine.core.DopeDoc
import ai.dopamine.core.DopeResolveInput
import ai.dopamine.core.DopeValue
import ai.dopamine.core.EffectRegistry
import ai.dopamine.core.number
import ai.dopamine.core.parseDope
import ai.dopamine.core.resolveDopeParams
import ai.dopamine.gl.DrawableEffect
import ai.dopamine.gl.EffectContext
import ai.dopamine.gl.EffectInstance
import ai.dopamine.gl.PassConfig
import ai.dopamine.gl.createPassInstance
import android.content.Context

/**
 * The single source of truth for the piece cap: BOTH the GLSL `#define MAX_PIECES`
 * and the integer-clamp const the `.dope` references (`clampMax: "MAX_PIECES"`).
 */
const val MAX_PIECES: Double = 120.0

class Confetti(context: Context) : DrawableEffect {
    override val name: String = "confetti"

    // Load the bundled `.dope` (the EXACT web bytes) from the merged APK assets —
    // proving the data spine is shared verbatim across platforms.
    val doc: DopeDoc = parseDope(
        context.assets.open("confetti.dope.json").bufferedReader(Charsets.UTF_8).use { it.readText() },
    )

    override fun resolve(feeling: DopeResolveInput): Map<String, DopeValue> =
        // `MAX_PIECES` is the only const; `pieceSeed` is the scatter key — both
        // byte-identical to the web `resolveDopeParams(...)` call.
        resolveDopeParams(doc, feeling, consts = mapOf("MAX_PIECES" to MAX_PIECES), scatterKey = "pieceSeed")

    override fun create(params: Map<String, DopeValue>, ctx: EffectContext): EffectInstance =
        createPassInstance(CONFIG, params, ctx)

    override val reducedMotionPeakMs: Double = 320.0
    override val reducedMotionHoldMs: Double = 420.0

    companion object {
        /** Construct + register the effect (needs a Context for the `.dope` asset). */
        fun register(context: Context): Confetti {
            val fx = Confetti(context.applicationContext)
            EffectRegistry.register(fx)
            return fx
        }

        // The launched + falling cloud spans a good chunk of the viewport; give the
        // shadow a moderate occluder "height". Matches the web SHADOW_HEIGHT_FRAC.
        private const val SHADOW_HEIGHT_FRAC: Double = 0.5

        private val CONFIG = PassConfig(
            vertex = CONFETTI_VERTEX_SRC,
            fragment = CONFETTI_FRAGMENT_SRC,
            uniforms = listOf(
                "uExposure", "uPieceCount", "uSpread", "uLaunchSpeed", "uGravity",
                "uFlutter", "uPieceSize", "uSpin", "uPieceSeed",
            ),
            usesOrigin = true,
            // overshoot feeds the envelope, not a uniform; pieceSeed is the scatter offset.
            bindings = mapOf("overshoot" to null, "pieceSeed" to "uPieceSeed"),
            shadowHeightFrac = { SHADOW_HEIGHT_FRAC },
            frame = { info, params ->
                // The launch-then-fall amplitude `confettiAmp` is Confetti's bespoke
                // timing — a sharp POP attack (overshoot at launch), a long luminous
                // sustain through the fall, then a soft fade as the last pieces settle.
                mapOf("amp" to confettiAmp(info.life, params.number("overshoot", 1.0)).toFloat())
            },
        )
    }
}
